client/me/connected-applications/index.jsx

```diff
@@ -106,79 +51,72 @@ const ConnectedApplications = createReactClass( {
-		const placeholders = [];
-
-		for ( let i = 0; i < 5; i++ ) {
-			placeholders.push(
-				<ConnectedAppItem
-					connection={ {
-						ID: i,
-						title: this.props.translate( 'Loading Connected Applications' ),
-					} }
-					key={ i }
-					isPlaceholder
-				/>
-			);
-		}
+	}
+
+	renderPlaceholders() {
+		const { translate } = this.props;
+
+		return times( 5, index => (
+			<ConnectedAppItem
+				connection={ {
+					ID: index,
+					title: translate( 'Loading Connected Applications' ),
+				} }
+				key={ index }
+				isPlaceholder
+			/>
+		) );
+	}
 
-		return placeholders;
-	},
-
-	renderConnectedApps: function() {
-		return this.props.connectedAppsData.initialized
-			? this.props.connectedAppsData.get().map( function( connection ) {
-					return (
-						<ConnectedAppItem
-							connection={ connection }
-							key={ connection.ID }
-							connectedApplications={ this.props.connectedAppsData }
-							revoke={ this.revokeConnection }
-						/>
-					);
-				}, this )
-			: this.renderPlaceholders();
-	},
-
-	renderConnectedAppsList: function() {
-		let connectedApps;
-		const hasConnectedApps = this.props.connectedAppsData.get().length;
-
-		if ( this.props.connectedAppsData.initialized ) {
-			connectedApps = hasConnectedApps ? this.renderConnectedApps() : this.renderEmptyContent();
-		} else {
-			connectedApps = this.renderConnectedApps();
+	renderConnectedApps() {
+		const { apps } = this.props;
+
+		if ( ! apps.length ) {
+			return this.renderPlaceholders();
 		}
 
+		return apps.map( connection => (
+			<ConnectedAppItem connection={ connection } key={ connection.ID } />
+		) );
+	}
+
+	renderConnectedAppsList() {
+		const { apps, isRequestingApps, path } = this.props;
+
 		return (
-			<div>
-				<SecuritySectionNav path={ this.props.path } />
+			<Fragment>
+				<SecuritySectionNav path={ path } />
 
-				{ connectedApps }
-			</div>
+				{ ! isRequestingApps && ! apps.length
+					? this.renderEmptyContent()
+					: this.renderConnectedApps() }
+			</Fragment>
 		);
-	},
+	}
+
+	render() {
+		const { translate } = this.props;
 
-	render: function() {
 		return (
 			<Main className="connected-applications">
+				<QueryConnectedApplications />
+
 				<PageViewTracker
 					path="/me/security/connected-applications"
 					title="Me > Connected Applications"
 				/>
 				<ReauthRequired twoStepAuthorization={ twoStepAuthorization } />
 				<MeSidebarNavigation />
 
-				<DocumentHead title={ this.props.translate( 'Connected Applications' ) } />
+				<DocumentHead title={ translate( 'Connected Applications' ) } />
 
 				{ this.renderConnectedAppsList() }
 			</Main>
 		);
-	},
-} );
+	}
+}
 
-export default connect( null, dispatch => bindActionCreators( { successNotice }, dispatch ) )(
-	localize( ConnectedApplications )
-);
+export default connect( state => ( {
+	apps: getConnectedApplications( state ),
+	isRequestingApps: get( getRequest( state, requestConnectedApplications() ), 'isLoading', false ),
```

Comment:
This should be redundant, but the selector isn't designed that way right now.

`null` should represent our not-loaded/not-in-state status.
An empty or populated lists of applications represent loaded data.

`getConnectedApplications` defaults to `[]` right now. If we change the default to `null` we can simplify this component's connect.

Reply:
Well, this was intentional. We seem to use `null` for the "data is not there yet" cases a lot (I personally used to do that often), but we must admit that it's a little bit of a hack, and doesn't always do what we want. For example, sometimes there might be more states than just requesting or not requesting, and the `null` wouldn't be enough for that. Also, if we return a `null` in the selector, and network requests consistently fail, we'll keep displaying a placeholder/loading state forever; while with this approach we'll be in the loading state only while the request is in flight.

Ideally, we'd like to leave all of that work to the data layer, which we are doing right in this PR.

Something we could improve in the future could be a HoC or some sort of trigger that would allow us to easily get access to that value without the need to use `get( getRequest( actionCreator() ) )`, but that's definitely something we could iterate on in the future.

Reply:
> We seem to use `null` for the "data is not there yet" […], but we must admit that it's a little bit of a hack

Using `null` as a value to indicate absence of data doesn't seem like a hack to me.

> there might be more states than just requesting or not requesting, and the `null` wouldn't be enough for that. If […] network requests consistently fail, we'll keep displaying a placeholder/loading state forever; while with this approach we'll be in the loading state only while the request is in flight

What is more relevant?

As it is now we provide (potentially incorrect) data and 2 _request_ states independently. The error case (bad/failed network request) is hidden and a potentially inaccurate view is shown allowing the user to interact with it.

In the other case, we represent two integrated _data_ states, `null` (no data) and presence of (as accurate as we can hope for) data. The error case remains unhandled at the component view level.

A different perspective: If we're happy with the user interacting with `[]` as a default application state, why even bother with placeholders? Let's just load with `[]` and inject the state when (and if) it arrives.

> Something we could improve in the future could be a HoC or some sort of trigger that would allow us to easily get access to that value without the need to use `get( getRequest( actionCreator() ) )`, but that's definitely something we could iterate on in the future.

I'm pretty sure this was a design decision to discourage request inspection. Maybe @dmsnell could provide more background?

Reply:
Thanks for taking the time to explain - I understand the rationale; it makes sense and am happy to iterate. I think it will indeed simplify the code in that case. 

Will follow-up in another PR with that.

Reply:
@sirreal - #25278 to the rescue

Reply:
I can say that `null` as a null value is perfectly normal and even encouraged. It means literally that we don't have the data - a typical tri-state value where `null` means we don't know, some empty value means we have loaded the data and nothing is there, and other data means we have loaded and something _is_ there.

We _do_ discourage interrogating individual network requests because usually we have better ways to represent the state assuming that "whatever is in state is the universal truth" but in those cases we have what you used and also I've been working on a beta version of a helper built upon the data layer that provides this intrinsically. Stay tuned!

Reply:
Thanks for chiming in @dmsnell! Looking forward to the new features 😎 

I've already suggested a PR to take advantage of the `null` and remove tracking of specific network requests, so that's landing soon.